Fix vcs password leak through subprocess_cmd.

a-feld opened the pull request:
pip installs that use auth with version control systems appear to be leaking passwords in the logs.

A partial fix appears to have been applied in #5773 but the passwords are still leaking due to `subprocess_cmd` printing the command description which includes the password in use.

This PR attempts to scrub the passwords out of the logs by using a new `HiddenText` class to hide the raw url contents in a `raw` attribute until explicitly needed.  The `subprocess_cmd` function has been modified to extract the `raw` attribute when calling the underlying command.

cjerdonek commented:
Thanks for filing this! I looked at the PR, and I think there is a cleaner way we can be handling this. Making every caller have to construct and pass `command_desc` and also track the `url_idx` seems more verbose and error-prone to me.

I'm thinking a simpler solution could involve defining a simple `str` wrapper class whose `__str__()` method defaults to the redacted form. That way it can be used in log formatting as is, and there is less chance of accidentally exposing it somewhere. The original, unredacted form could be accessed via an attribute on the class instance (e.g. `raw`), so it's opt-in for the sensitive form. 

The wrapper class could be something like this (the name is just a suggestion, maybe `HiddenText`):

```python
class SensitiveText(object):  

    def __init__(self, raw, redacted):
        self.raw = raw
        self.redacted = redacted
    
    def __str__(self):
        return self.redacted
```

And then we could have helper functions as needed so callers won't need to call `SensitiveText` themselves. Something like (again, names are just suggestions):

```python
def hide_url(url):
    redacted = redact_password_from_url(url)
    return SensitiveText(url, redacted=redacted)

hidden_url = hide_url(...)
```

and the even simpler:

```python
def hide_password(password):
    return SensitiveText(password, redacted='****')
```

Callers could "hide" variables as soon as we know about them (though for now, we could just restrict this use to the `vcs` modules).

Then `format_command_args()` could do an instance check for `SensitiveText` when constructing the command description `command_desc`: https://github.com/pypa/pip/blob/908e2034fa6b119bfc1a505c9f83140c1d2022a2/src/pip/_internal/utils/misc.py#L767-L772

And of course, we'd also want to convert these objects to `str` when actually invoking `subprocess`.

chrahunt commented:
@cjerdonek, that would also be relevant to #6834.

a-feld commented:
@cjerdonek Thanks for the feedback! I agree, the current approach in this PR feels pretty fragile. I'll start working on incorporating those suggestions!

a-feld commented:
@cjerdonek I've pushed up some edits, things are looking much cleaner code-wise now! 😄 
I haven't fully had the opportunity to vet these changes - I'm going to let CI run to see how much damage has been done but I figured in the meantime you might want to have a look 👀

cjerdonek commented on src/pip/_internal/utils/misc.py:
```diff
@@ -953,6 +963,36 @@ def _make_build_dir(build_dir):
     write_delete_marker_file(build_dir)
 
 
+class SensitiveText(str):
```
My instinct is that this class can be made a lot simpler. Like, couldn't it work just to do this? (no need for `__new__`, `__init__`, `__eq__`, and `__hash__`)

```python
class HiddenText(str):

    def __repr__(self):
        return '<HiddenText {!r}>'.format(str(self))

    def __str__(self):
        return self.redacted

    @property
    def raw(self):
        return str.__str__(self)
```

Then "public" creation functions can be made like this, for example:

```python
def hide_password(password):
    hidden = HiddenText(password)
    hidden.redacted = '****'

    return hidden

a = hide_password('*secret*')

print(str(a))
print(repr(a))
print(a.raw)
```

which gives:

```
****
<HiddenText '****'>
*secret*
```

I think decoupling the creation functions from the class will be better because it keeps the class simpler. Also, after giving it some thought, I think `HiddenText` is a better name in part because it's simpler to type and say.

By the way, your `__hash__()` implementation below I don't think is right because it bases the hash off of the redacted form, which can be the same for different values. The above, on the other hand, does the right thing because it inherits what `str` does, which is to use the underlying string value.

a-feld replied:
> the hash off of the redacted form

That's actually intentional. The idea is that if the redacted form is stored in a hashable location, the redacted form is likely what is being operated upon. The example that I ran into was the urlparse cache where the redacted url was being split. In that case (and I think all cases), the redacted form was what was being cached.

a-feld replied:
Also, the `__eq__` is necessary for proper comparison specifically during tests. I was trying to avoid rewriting all of the assertions which compare actual values to expected values but I'm open to rewriting those tests as well if that's a better approach.

cjerdonek replied:
> Also, the __eq__ is necessary for proper comparison specifically during tests.

Can you give me an example of some tests that would be affected by what I suggested?

a-feld replied:
This went back to when I was patching the `Link` model but at that point there were a few tests that were failing when comparing the actual url (HiddenText) against the expected url (str).

Here were a couple of examples:
https://github.com/pypa/pip/blob/908e2034fa6b119bfc1a505c9f83140c1d2022a2/tests/unit/test_req_file.py#L564

https://github.com/pypa/pip/blob/908e2034fa6b119bfc1a505c9f83140c1d2022a2/tests/unit/test_req_file.py#L535

Since I'm no longer updating the `Link` model, these tests wouldn't be impacted but if the intent is to move the `HiddenText` cast back into the `Link` model eventually this kind of issue will come up. At that point, it's whether the tests should be updated or the `__eq__` comparison logic.

I defer to your preferences here 😄

a-feld commented:
@cjerdonek I've pushed simplifications and renames to the `HiddenText` class. My concerns with the PR are more functional at the moment: I've edited the `Link` model to replace the `url` attribute with a `HiddenText` type. I'm not familiar enough with the overall architecture of pip to tell if I've actually broken something by doing this. Do you have any concerns with modifying the `Link` class? Should I move the `HiddenText` cast closer in to the `versioncontrol` logic?

a-feld commented:
@cjerdonek I'll probably also need some help with mypy due to the simplifications. Without a new/init, I think mypy is unable to detect the attributes which are set.

cjerdonek commented:
I think you should keep this initial PR more limited in scope, so don’t
change the Link class.

On Sun, Aug 11, 2019 at 9:23 PM Allan Feldman <notifications@github.com>
wrote:

> @cjerdonek <https://github.com/cjerdonek> I've pushed simplifications and
> renames to the HiddenText class. My concerns with the PR are more
> functional at the moment: I've edited the Link model to replace the url
> attribute with a HiddenText type. I'm not familiar enough with the
> overall architecture of pip to tell if I've actually broken something by
> doing this. Do you have any concerns with modifying the Link class?
> Should I move the HiddenText cast closer in to the versioncontrol logic?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pypa/pip/pull/6862?email_source=notifications&email_token=AACW33TTDUSCNM6PLHYQUS3QEDQUTA5CNFSM4IK44LJKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4BRH6A#issuecomment-520295416>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACW33VZS3OHKLEXDHKZCFLQEDQUTANCNFSM4IK44LJA>
> .
>

a-feld commented:
Scoping down the PR sounds like a good plan 😄
I've moved the `HiddenText` logic closer in to the version control logic!

cjerdonek commented:
> I've pushed simplifications and renames to the HiddenText class.

I still see the class methods..

a-feld commented:
> I still see the class methods.

Sorry, rereading the suggestion posted, I'm not entirely clear on the suggestion. Is the suggestion to remove the class methods entirely and replace with helper functions? Is there an advantage to having functions here over class methods? I'll go with whichever you suggest I'm just curious to learn more about the rationale.

cjerdonek commented:
Yes, the suggestion was to have the creation functions be separate from the class. This keeps the class simpler so it doesn't e.g. have to know about passwords, parsing URL's, etc. Also, if we later want to add something like a `HiddenUrl` class, we don't have to go around and change every call site. It's just in the one helper function.

a-feld commented:
@cjerdonek Got it, I also see what you were saying about the simplification. I'm making edits to be more in-line with your comment now 😄

a-feld commented:
@cjerdonek It doesn't look like the `HiddenText` class suggested [here](https://github.com/pypa/pip/pull/6862#discussion_r312775991) works. I'm not quite sure how to proceed. There are a couple of ways I can get this to work:

1. I can use the `HiddenText` class with the redacted value as the string and the raw value as an additional attribute. The downside to this is the eq behavior may have to be changed, depending on how we want to handle things.
2. I can remove the string subclass and pass the raw / redacted explicitly but this has the downside of not being able to work in areas that require strings or byte strings (like regex).
3. Possibly other ways 😄

a-feld commented:
@cjerdonek I've gone ahead with option 1 and left out all of the `__eq__` overrides / etc for now but the tests are failing as a result. The tests will have to be updated or the class will need an `__eq__` override to pass. Whichever is preferred I'll implement.

cjerdonek commented:
Can you say why it didn’t work for you?

On Sun, Aug 11, 2019 at 10:23 PM Allan Feldman <notifications@github.com>
wrote:

> @cjerdonek <https://github.com/cjerdonek> It doesn't look like the
> HiddenText class suggested here
> <https://github.com/pypa/pip/pull/6862#discussion_r312775991> works. I'm
> not quite sure how to proceed. There are a couple of ways I can get this to
> work:
>
>    1. I can use the HiddenText class with the redacted value as the
>    string and the raw value as an additional attribute. The downside to this
>    is the eq behavior may have to be changed, depending on how we want to
>    handle things.
>    2. I can remove the string subclass and pass the raw / redacted
>    explicitly but this has the downside of not being able to work in areas
>    that require strings or byte strings (like regex).
>    3. Possibly other ways 😄
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pypa/pip/pull/6862?email_source=notifications&email_token=AACW33SBLGUEDFVY6I2YTKLQEDXTXA5CNFSM4IK44LJKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4BS3SA#issuecomment-520302024>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACW33UDZL73LTXQY5WZ7WDQEDXTXANCNFSM4IK44LJA>
> .
>

a-feld commented:
> Can you say why it didn’t work for you?

When I used the class given, I observed the passwords getting leaked into the logs. This was due to the call to [format_command_args](https://github.com/pypa/pip/blob/908e2034fa6b119bfc1a505c9f83140c1d2022a2/src/pip/_internal/utils/misc.py#L875) which was ultimately using the raw value to formulate the description.

cjerdonek commented:
> When I used the class given, I observed the passwords getting leaked into the logs. This was due to the call to format_command_args which was ultimately using the raw value to formulate the description.

This is why I said this in my first comment: 

> Then format_command_args() could do an instance check for SensitiveText when constructing the command description command_desc:

a-feld commented:
I worry that defaulting to operating on secrets is still error prone. Will there be future issues that occur from defaulting to the use of the raw secret value for string operations? The current implementation is fail-secure, the raw value must be explicitly accessed. I can certainly see advantages / disadvantages either way, it does seem prudent to discuss though.

cjerdonek commented:
I would focus on getting something working that is simple and doesn't require too many changes. And we can always continue to refine. My suggestion in my first comment was "fail secure" in the sense that you're talking about, but I don't know what affect it would have on type-checking without trying it out myself. Personally, I'd rather we didn't subclass from `str` because it is inherently more complex and harder to reason about (and has many methods we have to consider, etc).

a-feld commented:
I pushed up an edit which does an isinstance check inside of `format_command_args`. Thanks for taking the time spent so far to help me out on this PR @cjerdonek !

cjerdonek commented on src/pip/_internal/utils/misc.py:
```diff
@@ -769,7 +769,20 @@ def format_command_args(args):
     """
-    return ' '.join(shlex_quote(arg) for arg in args)
+    return ' '.join(
+        shlex_quote(arg.redacted) if isinstance(
```
This can just be `shlex_quote(str(arg))`.

cjerdonek commented:
> I pushed up an edit which does an isinstance check inside of format_command_args. 

It looks like the `isinstance()` check isn't there anymore. The reason I would keep that check is because we can't rely on the type-checker 100%. Like, in Python 2, it's possible that some of the arguments might have `unicode` type (I recall one function annotated with `str` that can return a unicode path, but it has an `ignore` comment), and if there's a non-ascii character, then passing it to `str()` can cause a `UnicodeEncodeError`.

cjerdonek commented on src/pip/_internal/download.py:
```diff
@@ -722,7 +723,8 @@ def is_archive_file(name):
 def unpack_vcs_link(link, location):
     vcs_backend = _get_used_vcs_backend(link)
-    vcs_backend.unpack(location, url=link.url)
+    url = hide_url(link.url)
```
Even though this seems harmless, I think it would be better to keep things narrow for now and just keep to the minimal places in the vcs code. We can expand the use in later PR's as the pattern solidifies, etc.

cjerdonek commented on src/pip/_internal/utils/misc.py:
```diff
@@ -953,6 +962,32 @@ def _make_build_dir(build_dir):
     write_delete_marker_file(build_dir)
 
 
+class HiddenText(str):
```
You should add type annotations to all the new methods and functions you're adding.

cjerdonek commented on src/pip/_internal/utils/misc.py:
```diff
@@ -953,6 +962,32 @@ def _make_build_dir(build_dir):
+        return str.__str__(self)
+
+
+def hide_url(raw):
```
It would be good to add unit tests of this function (and the other) testing some basic properties of the return value (e.g. `repr()`, `str()`, the attribute access, etc.).

cjerdonek commented on tests/functional/test_install.py:
```diff
@@ -1527,3 +1527,46 @@ def test_target_install_ignores_distutils_config_install_prefix(script):
     ), str(result)
+
+
+@pytest.mark.network
```
It would be best to avoid adding more tests that depend on network access, IMO, as it adds flakiness, makes the tests slower, etc. It would be best if we can find the appropriate places at which to unit test to test this functionality (starting with the "inner-most" functions).

chrahunt commented on src/pip/_internal/utils/misc.py:
```diff
@@ -765,15 +765,33 @@ def unpack_file(
+    # Casting to a str when a non-ascii character is present and the type(arg)
+    # is unicode can trigger a UnicodeDecodeError.
+    return ' '.join(
+        shlex_quote(str(arg)) if isinstance(
```
I think the prior comment is still pending, can be `shlex_quote(str(arg)) for arg in args`.

a-feld replied:
@chrahunt Thanks for the feedback, I actually originally edited this to be `shlex_quote(str(arg)) for arg in args`.

@cjerdonek [requested](https://github.com/pypa/pip/pull/6862#issuecomment-520327996) here that I change to use `isinstance`.

I tried to create a `UnicodeDecodeError` appear in this function using py27 but was unable to. I think it's ok to program defensively. I also think this is unlikely to fail in actual operation. 😄 

Based on the feedback though, I'm not sure if I should leave the `isinstance` check in or take it out

cjerdonek replied:
I'd rather not risk introducing errors that we don't need to risk.

chrahunt replied:
Thanks, sorry I had missed that comment.

a-feld commented:
@cjerdonek I've tried to address all of the feedback points. Apologies in advance if I omitted something. 🙏 

1. converted to using a HiddenText class that does not subclass from str (there were many changes to logic and type annotations, hence the large diff)
2. Added unit tests for the `hide_*` helpers
3. I've removed all the integration tests as requested.

For the testing parts of things:
> It would be best to avoid adding more tests that depend on network access
I actually realized that I didn't _strictly_ depend on the network being available. The tests were only checking to see if the password ended up in the logs. I'm not sure if you had ideas for implementing a regression test that is also performant that would have caught the password logging issue. I'd love to have an end-to-end test that would catch these kind of leaks! 🐛

cjerdonek commented:
> converted to using a HiddenText class that does not subclass from str (there were many changes to logic and type annotations, hence the large diff)

My original suggestion was in response to your first PR, which was simpler and tried to do less. I was suggesting a replacement for your `url_idx` approach, etc. But since then, it seems like the PR has been expanding in scope, which is why the diff is now so big. If you go back to the approach of your first PR and then do what I was suggesting, I think things will be simpler than what you first proposed. Basically, I was suggesting replacing things like:

```python
cls.run_command(
    ['fetch', '-q', url] + rev_options.to_args(),
    cwd=dest,
)
```

with 

```python
cls.run_command(
    ['fetch', '-q', hide_url(url)] + rev_options.to_args(),
    cwd=dest,
)
```

and then, once you update the type annotations and make a few other minor changes, I think everything should work to address the issue this is meant to fix (the `command_desc` leaking auth info).

cjerdonek commented on src/pip/_internal/utils/misc.py:
```diff
@@ -765,15 +765,33 @@ def unpack_file(
 
 def format_command_args(args):
-    # type: (List[str]) -> str
+    # type: (Union[List[str], List[Union[str, HiddenText]]]) -> str
```
It looks like you have one too many `Union`s. Can't this just be `List[Union[str, HiddenText]]`?

a-feld replied:
mypy doesn't allow it because this function is in use by [req_install.py](https://github.com/pypa/pip/blob/76a89542a8ea8b9ebbfe393f23ccbe764c727d3e/src/pip/_internal/req/req_install.py#L544) and a few other files.

mypy appears to treat `List[Union[str, HiddenText]]` as incompatible with `List[str]`.

cjerdonek replied:
Okay, you might want to define a short-hand for `List[Union[str, HiddenText]]` then to make the type annotations more concise. `CommandArgs`? Then it can be `Union[List[str], CommandArgs]`.

a-feld commented:
@cjerdonek Ah, I see! Sorry for the misunderstanding - I read the original feedback as there being a way to avoid intercepting all of the `run_command` points by intercepting the URL upstream using a special type based on this comment:

> Callers could "hide" variables as soon as we know about them

I thought it was a cool idea and tried playing around with it! I've pushed an update going back to the original plan of patching all of the URL locations in each of the vcs modules.

cjerdonek commented:
> I thought it was a cool idea and tried playing around with it! 

I agree, and thanks for your efforts trying it out! :) I think we both got a taste for what that might involve. Since that "next step" would take more work and planning, etc. to make sure it all works out (as it's more extensive and potentially more sophisticated), I think it's better for now to save something like that for a later PR, and just get something simpler in place as a start to address the immediate issue..

a-feld commented:
@cjerdonek 👍 ! Any ideas on regression tests for this? If I remove the network marker from the tests submitted earlier in the PR, will those be ok? They do tend to run on the slower side.

cjerdonek commented on src/pip/_internal/vcs/bazaar.py:
```diff
@@ -41,23 +46,26 @@ def export(self, location, url):
             show_stdout=False,
         )
 
     def fetch_new(self, dest, url, rev_options):
```
It looks like the PR would become simpler if you annotate some of the methods like these so that the `url` argument has to be `HiddenText`. (But not all. Methods like `get_url_rev_options()` and `resolve_revision()` you probably want to leave as accepting only `str`.) Then many of the function bodies won't need to change, and you can lean on the type-checker as opposed to tests for a lot of the correctness.

a-feld commented:
@cjerdonek OK I've pushed up an edit which annotates the inputs to the VCS methods and hides the url in versioncontrol.py

The only known deviation is that `resolve_revision` in git uses HiddenText for the url since a `git fetch ... $url` command is run which can leak passwords if not hidden.

 Hopefully I've addressed all feedback 🙏

cjerdonek commented:
One thing that can be done to simplify the changes would be to define a function like this and put it in the same module that you define `CommandArgs`:

```python
def cast_args(args):
    # type: (Union[List[str], List[HiddenText], CommandArgs]) -> CommandArgs
    new_args = []  # type: CommandArgs
    new_args.extend(args)

    return new_args
```

Then instead of having to break every list up into several lines, you can just convert the first element. So things like this:

```python
cmd_args = (
    ['export'] + self.get_remote_call_options() +
    rev_options.to_args() + [url, location]
)
```

would just need to be changed to:

```python
cmd_args = (
    cast_args(['export']) + self.get_remote_call_options() +
    rev_options.to_args() + [url, location]
)
```

Also, question: how many changes to the code were due just to adding the type annotations themselves and not related to `HiddenText`? I'm wondering if it would make sense to add type annotations to the vcs modules without `HiddentText` in a separate PR, so those code changes can be looked at independently.

a-feld commented:
Aha! I can use a type cast helper! Thanks for the tip, I'm not very practiced with python type annotations 😄 

> how many changes to the code were due just to adding the type annotations themselves and not related to HiddenText?

I tried to make all of the type annotation updates in a specific commit to hopefully ease some of the review pain
https://github.com/pypa/pip/pull/6862/commits/eff94f71a47370a307ccc6f96a4094de8a3caf04

There should be a way in github to filter by specific commit when reviewing the diffs as well.

cjerdonek commented:
Would you be open to filing a PR that adds type annotations to the vcs
modules without any changes related to HiddenText? That would let us see
that CI is passing. Also, in your PR I saw some changes that didn’t seem
related to the hidden text stuff (e.g. around None values), so I’m assuming
it was done to get the type checker to pass.



On Tue, Aug 13, 2019 at 7:08 PM Allan Feldman <notifications@github.com>
wrote:

> Aha! I can use a type cast helper! Thanks for the tip, I'm not very
> practiced with python type annotations 😄
>
> how many changes to the code were due just to adding the type annotations
> themselves and not related to HiddenText?
>
> I tried to make all of the type annotation updates in a specific commit to
> hopefully ease some of the review pain
> eff94f7
> <https://github.com/pypa/pip/commit/eff94f71a47370a307ccc6f96a4094de8a3caf04>
>
> There should be a way in github to filter by specific commit when
> reviewing the diffs as well.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pypa/pip/pull/6862?email_source=notifications&email_token=AACW33XV44RN3ZM5T53OIWDQENSJTA5CNFSM4IK44LJKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4HPRYI#issuecomment-521074913>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACW33QNRGLROXNLQPRMAQTQENSJTANCNFSM4IK44LJA>
> .
>

a-feld commented:
> Would you be open to filing a PR that adds type annotations to the vcs modules without any changes related to HiddenText?

If you're ok opening and driving that second PR, we can merge that one first and then I can rebase on top of those changes. To be honest I'm having trouble mentally tracking all of the changes flying around at the moment and I feel that driving another PR would add to my cognitive overhead.

> I saw some changes that didn’t seem related to the hidden text stuff (e.g. around None values), so I’m assuming it was done to get the type checker to pass.

Yeah, that's correct. In particular the none check you're referring to is because the property returns an optional so the type checker was complaining about accessing an attribute on None-typed values.

cjerdonek commented:
You would be able to set this one down while doing that one. The purpose
would be to make each PR do fewer things. Otherwise, I would recommend you
go back to the PR prior to when I suggested that you change the argument
type, because I had assumed the vcs modules were already annotated.

On Tue, Aug 13, 2019 at 7:25 PM Allan Feldman <notifications@github.com>
wrote:

> Would you be open to filing a PR that adds type annotations to the vcs
> modules without any changes related to HiddenText?
>
> If you're ok opening and driving that second PR, we can merge that one
> first and then I can rebase on top of those changes. To be honest I'm
> having trouble mentally tracking all of the changes flying around at the
> moment and I feel that driving another PR would add to my cognitive
> overhead.
>
> I saw some changes that didn’t seem related to the hidden text stuff (e.g.
> around None values), so I’m assuming it was done to get the type checker to
> pass.
>
> Yeah, that's correct. In particular the none check you're referring to is
> because the property returns an optional so the type checker was
> complaining about accessing an attribute on None-typed values.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pypa/pip/pull/6862?email_source=notifications&email_token=AACW33QDQ5ATKYVUW23EVQTQENUJTA5CNFSM4IK44LJKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4HQIUY#issuecomment-521077843>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACW33UVOB2KCWFRZCVYX2TQENUJTANCNFSM4IK44LJA>
> .
>

a-feld commented:
👍  I'll open a 2nd PR for the type annotations and put this one on hold for now.

cjerdonek commented:
I posted PR #6890 to finish the work started here, FYI.
